addons/l10n_in_edi_ewaybill/models/account_edi_format.py

```diff
@@ -0,0 +1,601 @@
+                authenticate_response = self._l10n_in_edi_authenticate(invoices.company_id)
+                if not authenticate_response.get("error"):
+                    error = []
+                    response = self._l10n_in_edi_irn_ewaybill_generate(invoices.company_id, generate_json)
+                    if response.get("error"):
+                        error = response["error"]
+                        error_codes = [e.get("code") for e in error]
+            if "4002" in error_codes:
+                # Get E-waybill by details in case of IRN is already generated
+                # this happens when timeout from the Government portal but E-waybill is generated
+                response = self._l10n_in_edi_irn_ewaybill_get(invoices.company_id, generate_json.get("Irn"))
+                if not response.get("error"):
+                    error = []
+                    odoobot = self.env.ref("base.partner_root")
+                    invoices.message_post(author_id=odoobot.id, body=_(
+                        "Somehow this E-waybill is generate in government portal before." \
+                        "<br/>Normally, this should not happen too often" \
+                        "<br/>Just verify value by enter details into government website " \
+                        "<a href='https://ewaybillgst.gov.in/Others/EBPrintnew.aspx'>here<a>."
+                    ))
+            if "no-credit" in error_codes:
+                res[invoices] = {
+                    "success": False,
+                    "error": self._l10n_in_edi_get_iap_buy_credits_message(invoices.company_id),
+                    "blocking_level": "error",
+                }
+            elif error:
+                error_message = "<br/>".join(["[%s] %s" % (e.get("code"), html_escape(e.get("message"))) for e in error])
+                blocking_level = "error"
+                if "404" in error_codes or "waiting" in error_codes:
+                    blocking_level = "warning"
+                res[invoices] = {
+                    "success": False,
+                    "error": error_message,
+                    "blocking_level": blocking_level,
+                }
+        if not response.get("error"):
+            json_dump = json.dumps(response.get("data"))
+            json_name = "%s_irn_ewaybill.json" % (invoices.name.replace("/", "_"))
+            attachment = self.env["ir.attachment"].create({
+                "name": json_name,
+                "raw": json_dump.encode(),
+                "res_model": "account.move",
+                "res_id": invoices.id,
+                "mimetype": "application/json",
+            })
+            inv_res = {"success": True, "attachment": attachment}
+            res[invoices] = inv_res
+        return res
+
+    def _l10n_in_edi_irn_ewaybill_generate_json(self, invoice):
+        json_payload = {
+            "Irn": invoice._get_l10n_in_edi_response_json().get("Irn"),
```

Comment:
should we still generate it without irn?

Reply:
yes, only if your invoice is not applicable for E-invoice or the company is not eligible for E-invoice